gQuant/plugins/cusignal_plugin/greenflow_cusignal_plugin/convolution/fftconvolve.py

```diff
@@ -0,0 +1,127 @@
+A string indicating the size of the output:
+``full``
+   The output is the full discrete linear convolution
+   of the inputs. (Default)
+``valid``
+   The output consists only of those elements that do not
+   rely on the zero-padding. In 'valid' mode, either `in1` or `in2`
+   must be at least as large as the other in every dimension.
+``same``
+   The output is the same size as `in1`, centered
+   with respect to the 'full' output.
+   axis : tuple, optional
+'''
+
+_FFTCONV_AXES_DESC = '''axes : int or array_like of ints or None, optional
+Axes over which to compute the convolution.
+The default is over all axes.
+'''
+
+
+class CusignalFFTConvolveNode(TemplateNodeMixin, Node):
+    def init(self):
+        TemplateNodeMixin.init(self)
+        port_type = PortsSpecSchema.port_type
+        inports = {
+            'in1': {port_type: [cp.ndarray, np.ndarray]},
+            'in2': {port_type: [cp.ndarray, np.ndarray]}
+        }
+        outports = {
+            'fftconvolve': {port_type: [cp.ndarray, np.ndarray]},
+        }
+        self.template_ports_setup(in_ports=inports, out_ports=outports)
+
+        meta_outports = {'fftconvolve': {}}
+        self.template_meta_setup(out_ports=meta_outports)
+
+    def conf_schema(self):
+        mode_enum = ['full', 'valid', 'same']
+        json = {
+            'title': 'Cusignal Convolution Node',
+            'type': 'object',
+            'description': _FFTCONV_DESC,
+            'properties': {
+                'mode': {
+                    'type': 'string',
+                    'description': _FFTCONV_MODE_DESC,
+                    'enum': mode_enum,
+                    'default': 'full'
+                },
+                'axes': {
+                    'type': 'array',
+                    'items': {
+                        'type': 'integer'
+                    },
+                    'description': _FFTCONV_AXES_DESC,
+                },
+                'use_cpu': {
+                    'type': 'boolean',
+                    'description': 'Use CPU for computation via '
+                        'scipy::signal.fftconvolve. Default is False and '  # noqa: E131,E501
+                        'runs on GPU via cusignal.',
+                    'default': False
+                },
+            },
+        }
+        return ConfSchema(json=json)
+
+    def process(self, inputs):
+        mode = self.conf.get('mode', 'full')
+        axes = self.conf.get('axes', [])
+        use_cpu = self.conf.get('use_cpu', False)
+
+        in1 = inputs['in1']
+        in2 = inputs['in2']
+
+        if len(axes) == 0:
+            axes = None
+        elif len(axes) == 1:
+            axes = axes[0]
+
+        if use_cpu:
+            fftconv = sifftconv(in1, in2, mode=mode, axes=axes)
+        else:
+            cache = cp.fft.config.get_plan_cache()
+            cache.clear()
+            mempool = cp.get_default_memory_pool()
+            mempool.free_all_blocks()
+
```

Comment:
Yes, seems to be necessary. Cusignal's fftconvolve causes crazy GPU memory leaks. If you call cusignal's fftconvolve a couple of times in a row, it will use up all the GPU memory and crash. Messing with the cupy's fft config cache I was able to get it under control.

Reply:
okay